File: geo/services/ban.py
from __future__ import annotations

import asyncio

import httpx

from config import get_settings
from models import CitySuggestion


async def _search_ban(query: str, *, limit: int, ban_type: str | None = None) -> list[dict]:
    settings = get_settings()
    query = " ".join(query.split()).strip()
    if not query:
        raise ValueError("query is required")

    last_error: Exception | None = None

    async with httpx.AsyncClient(
        base_url=settings.ban_api_url,
        timeout=settings.http_timeout_seconds,
    ) as client:
        for attempt in range(1, settings.ban_retry_attempts + 1):
            try:
                response = await client.get(
                    "/search/",
                    params={
                        "q": query,
                        "limit": limit,
                        "autocomplete": 1,
                        **({"type": ban_type} if ban_type else {}),
                    },
                )
                response.raise_for_status()
                payload = response.json()
                features = payload.get("features") or []
                if not features:
                    raise LookupError(f"BAN returned no result for query: {query}")

                return features
            except (httpx.HTTPError, LookupError, ValueError) as exc:
                last_error = exc
                if attempt == settings.ban_retry_attempts:
                    break
                await asyncio.sleep(2 ** (attempt - 1))

    raise RuntimeError(f"failed to search BAN for query: {query}") from last_error


def _parse_coordinates(feature: dict, query: str) -> tuple[float, float]:
    coordinates = feature.get("geometry", {}).get("coordinates") or []
    if len(coordinates) < 2:
        raise LookupError(f"BAN returned invalid coordinates for: {query}")

    return float(coordinates[0]), float(coordinates[1])


async def resolve_address(address: str) -> tuple[float, float]:
    features = await _search_ban(address, limit=1)
    return _parse_coordinates(features[0], address)


async def resolve_city(city: str) -> tuple[float, float]:
    features = await _search_ban(city, limit=1, ban_type="municipality")
    return _parse_coordinates(features[0], city)


async def search_cities(query: str, *, limit: int = 5) -> list[CitySuggestion]:
    features = await _search_ban(query, limit=limit, ban_type="municipality")
    suggestions: list[CitySuggestion] = []

    for feature in features:
        lon, lat = _parse_coordinates(feature, query)
        properties = feature.get("properties", {})
        suggestions.append(
            CitySuggestion(
                name=properties.get("label", ""),
                city=properties.get("city", ""),
                postcode=properties.get("postcode", ""),
                lat=lat,
                lon=lon,
            )
        )

    return suggestions

File: back/internal/ingestion/geocoder/ban.go
package geocoder

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"time"

	"github.com/tanguyRa/onvaou/internal/config"
)

type BANClient struct {
	baseURL       string
	httpClient    *http.Client
	retryAttempts int
}

func NewBANClient(cfg config.Config, httpClient *http.Client) *BANClient {
	return &BANClient{
		baseURL:       cfg.Ingestion.BANAPIURL,
		httpClient:    httpClient,
		retryAttempts: cfg.Ingestion.BANRetryAttempts,
	}
}

func (c *BANClient) ResolveAddress(ctx context.Context, address string) (float64, float64, error) {
	return c.search(ctx, address, 1, "")
}

func (c *BANClient) ResolveCity(ctx context.Context, city string) (float64, float64, error) {
	return c.search(ctx, city, 1, "municipality")
}

func (c *BANClient) search(ctx context.Context, query string, limit int, banType string) (float64, float64, error) {
	if query == "" {
		return 0, 0, fmt.Errorf("query is required")
	}

	var lastErr error
	attempts := c.retryAttempts
	if attempts < 1 {
		attempts = 1
	}

	for attempt := 1; attempt <= attempts; attempt++ {
		lon, lat, err := c.doSearch(ctx, query, limit, banType)
		if err == nil {
			return lon, lat, nil
		}
		lastErr = err
		if attempt < attempts {
			select {
			case <-ctx.Done():
				return 0, 0, ctx.Err()
			case <-time.After(time.Duration(1<<(attempt-1)) * time.Second):
			}
		}
	}

	return 0, 0, fmt.Errorf("failed to search BAN for query %q: %w", query, lastErr)
}

func (c *BANClient) doSearch(ctx context.Context, query string, limit int, banType string) (float64, float64, error) {
	endpoint, err := url.Parse(c.baseURL)
	if err != nil {
		return 0, 0, err
	}
	endpoint.Path += "/search/"
	params := endpoint.Query()
	params.Set("q", query)
	params.Set("limit", fmt.Sprintf("%d", limit))
	params.Set("autocomplete", "1")
	if banType != "" {
		params.Set("type", banType)
	}
	endpoint.RawQuery = params.Encode()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint.String(), nil)
	if err != nil {
		return 0, 0, err
	}

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return 0, 0, err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		return 0, 0, fmt.Errorf("ban upstream returned status %d", resp.StatusCode)
	}

	var payload struct {
		Features []struct {
			Geometry struct {
				Coordinates []float64 `json:"coordinates"`
			} `json:"geometry"`
		} `json:"features"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&payload); err != nil {
		return 0, 0, err
	}
	if len(payload.Features) == 0 || len(payload.Features[0].Geometry.Coordinates) < 2 {
		return 0, 0, fmt.Errorf("ban returned no coordinates")
	}

	return payload.Features[0].Geometry.Coordinates[0], payload.Features[0].Geometry.Coordinates[1], nil
}
